File: dev-server.js
#!/usr/bin/env node

/**
 * Development Server with Live Reload
 * Watches for file changes and automatically refreshes the browser
 * Similar to VS Code's "Live Server" extension
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

// Configuration
const PORT = 8080;
const ROOT_DIR = path.join(__dirname, 'frontend');
const WATCH_EXTENSIONS = ['.html', '.css', '.js', '.json'];
const DEBOUNCE_DELAY = 100;

// MIME types
const MIME_TYPES = {
  '.html': 'text/html',
  '.css': 'text/css',
  '.js': 'application/javascript',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.pdf': 'application/pdf',
  '.csv': 'text/csv',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.eot': 'application/vnd.ms-fontobject'
};

// Store connected SSE clients
const sseClients = new Set();

// Live reload script to inject into HTML
const LIVE_RELOAD_SCRIPT = `
<script>
(function() {
  const source = new EventSource('/__live-reload__');
  let retryCount = 0;
  const maxRetries = 10;

  source.onmessage = function(e) {
    if (e.data === 'reload') {
      console.log('🔄 Reloading page due to file changes...');
      location.reload();
    }
  };

  source.onerror = function(e) {
    console.error('Live reload connection error:', e);
    retryCount++;

    if (retryCount > maxRetries) {
      console.error('Max retries reached. Live reload disabled.');
      source.close();
    } else {
      console.log(\`Retrying connection... (attempt \${retryCount}/\${maxRetries})\`);
    }
  };

  source.onopen = function() {
    retryCount = 0;
    console.log('✅ Live reload connected');
  };

  // Heartbeat to keep connection alive
  setInterval(() => {
    if (source.readyState === EventSource.OPEN) {
      // Connection is alive
    }
  }, 30000);
})();
</script>
`;

// Create HTTP server
const server = http.createServer((req, res) => {
  // Handle SSE endpoint for live reload
  if (req.url === '/__live-reload__') {
    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'Connection': 'keep-alive',
      'Access-Control-Allow-Origin': '*'
    });

    // Send initial connection message
    res.write('data: connected\n\n');

    // Add client to set
    sseClients.add(res);

    // Remove client on disconnect
    req.on('close', () => {
      sseClients.delete(res);
    });

    // Keep connection alive with periodic pings
    const keepAlive = setInterval(() => {
      res.write(':ping\n\n');
    }, 30000);

    req.on('close', () => {
      clearInterval(keepAlive);
    });

    return;
  }

  // Resolve file path
  let filePath = path.join(ROOT_DIR, req.url === '/' ? 'index.html' : req.url);

  // Security: prevent directory traversal
  if (!filePath.startsWith(ROOT_DIR)) {
    res.writeHead(403);
    res.end('Forbidden');
    return;
  }

  // Check if path is directory and serve index.html
  if (fs.existsSync(filePath) && fs.statSync(filePath).isDirectory()) {
    filePath = path.join(filePath, 'index.html');
  }

  // Serve file
  fs.readFile(filePath, (err, content) => {
    if (err) {
      if (err.code === 'ENOENT') {
        res.writeHead(404);
        res.end('404 Not Found');
      } else {
        res.writeHead(500);
        res.end(`Server Error: ${err.code}`);
      }
      return;
    }

    const ext = path.extname(filePath).toLowerCase();
    const mimeType = MIME_TYPES[ext] || 'application/octet-stream';

    // Inject live reload script into HTML files
    if (ext === '.html') {
      const htmlString = content.toString();
      const modifiedHtml = htmlString.replace('</body>', `${LIVE_RELOAD_SCRIPT}</body>`);
      content = Buffer.from(modifiedHtml);
    }

    res.writeHead(200, {
      'Content-Type': mimeType,
      'Cache-Control': 'no-cache, no-store, must-revalidate',
      'Access-Control-Allow-Origin': '*'
    });
    res.end(content);
  });
});

// File watcher with debouncing
let reloadTimeout;
const watchedDirs = [
  path.join(ROOT_DIR),
  path.join(ROOT_DIR, 'css'),
  path.join(ROOT_DIR, 'js'),
  path.join(ROOT_DIR, 'js', 'core'),
  path.join(ROOT_DIR, 'js', 'views'),
  path.join(ROOT_DIR, 'data', 'config')
];

function notifyClients() {
  sseClients.forEach(client => {
    client.write('data: reload\n\n');
  });
}

function setupWatchers() {
  console.log('📁 Watching directories:');
  watchedDirs.forEach(dir => {
    if (!fs.existsSync(dir)) {
      console.log(`  ⚠️  ${dir} (not found, skipping)`);
      return;
    }

    console.log(`  ✓ ${dir}`);

    fs.watch(dir, { recursive: false }, (eventType, filename) => {
      if (!filename) return;

      const ext = path.extname(filename).toLowerCase();
      if (!WATCH_EXTENSIONS.includes(ext)) return;

      // Debounce reload notifications
      clearTimeout(reloadTimeout);
      reloadTimeout = setTimeout(() => {
        console.log(`📝 File changed: ${filename}`);
        notifyClients();
      }, DEBOUNCE_DELAY);
    });
  });
}

// Start server
server.listen(PORT, () => {
  console.log('');
  console.log('🚀 VioBox Viewer Development Server');
  console.log('====================================');
  console.log(`📡 Server:     http://localhost:${PORT}`);
  console.log(`📂 Root:       ${ROOT_DIR}`);
  console.log(`🔄 Live Reload: Enabled`);
  console.log('');
  console.log('Press Ctrl+C to stop the server');
  console.log('');

  setupWatchers();

  // Try to open browser (works on most platforms)
  const url = `http://localhost:${PORT}`;
  const platform = process.platform;

  let command;
  if (platform === 'win32') {
    command = `start ${url}`;
  } else if (platform === 'darwin') {
    command = `open ${url}`;
  } else {
    command = `xdg-open ${url} 2>/dev/null || echo "Please open ${url} in your browser"`;
  }

  exec(command, (err) => {
    if (!err) {
      console.log('🌐 Browser opened automatically');
    }
  });
});

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('\n👋 Shutting down server...');

  // Close all SSE connections
  sseClients.forEach(client => {
    client.end();
  });

  server.close(() => {
    console.log('Server stopped');
    process.exit(0);
  });
});

// Handle errors
server.on('error', (err) => {
  if (err.code === 'EADDRINUSE') {
    console.error(`❌ Port ${PORT} is already in use!`);
    console.error('   Try one of these solutions:');
    console.error(`   1. Stop the other process using port ${PORT}`);
    console.error('   2. Change the PORT variable in this script');
    process.exit(1);
  } else {
    console.error('Server error:', err);
  }
});

File: dev-server.py
#!/usr/bin/env python3

"""
Development Server with Live Reload for VioBox Viewer
Watches for file changes and automatically refreshes the browser
Similar to VS Code's "Live Server" extension
"""

import os
import sys
import time
import json
import mimetypes
import threading
import webbrowser
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler
from datetime import datetime

# Configuration
PORT = 8080
ROOT_DIR = Path(__file__).parent / 'frontend'
WATCH_EXTENSIONS = {'.html', '.css', '.js', '.json'}
DEBOUNCE_DELAY = 0.5  # seconds

# Global state
sse_clients = []
last_reload_time = 0
reload_lock = threading.Lock()

# Live reload JavaScript to inject
LIVE_RELOAD_SCRIPT = """
<script>
(function() {
    const source = new EventSource('/__live_reload_sse__');
    let retryCount = 0;
    const maxRetries = 10;

    source.onmessage = function(e) {
        const data = JSON.parse(e.data);
        if (data.action === 'reload') {
            console.log('🔄 Reloading page due to changes in:', data.file);
            location.reload();
        }
    };

    source.onerror = function(e) {
        retryCount++;
        if (retryCount > maxRetries) {
            console.error('Live reload connection failed. Please refresh manually.');
            source.close();
        }
    };

    source.onopen = function() {
        retryCount = 0;
        console.log('✅ Live reload connected on port %d', source.url);
    };

    // Keep connection alive
    setInterval(() => {
        if (source.readyState === EventSource.CONNECTING) {
            console.log('Reconnecting to live reload...');
        }
    }, 30000);
})();
</script>
""" % PORT

class FileChangeHandler(FileSystemEventHandler):
    """Handles file system events and triggers reload"""

    def __init__(self, server):
        self.server = server

    def on_modified(self, event):
        if event.is_directory:
            return

        file_path = Path(event.src_path)
        if file_path.suffix.lower() in WATCH_EXTENSIONS:
            self.trigger_reload(file_path.name)

    def trigger_reload(self, filename):
        global last_reload_time

        current_time = time.time()
        with reload_lock:
            # Debounce rapid changes
            if current_time - last_reload_time < DEBOUNCE_DELAY:
                return
            last_reload_time = current_time

        print(f"📝 File changed: {filename}")

        # Notify all connected clients
        message = json.dumps({
            'action': 'reload',
            'file': filename,
            'timestamp': datetime.now().isoformat()
        })

        # Send reload message to all SSE clients
        for client in sse_clients[:]:  # Copy list to avoid modification during iteration
            try:
                client.wfile.write(f"data: {message}\n\n".encode())
                client.wfile.flush()
            except:
                # Remove disconnected clients
                sse_clients.remove(client)

class LiveReloadHTTPHandler(SimpleHTTPRequestHandler):
    """HTTP handler with live reload support"""

    def __init__(self, *args, **kwargs):
        # Change to frontend directory
        os.chdir(ROOT_DIR)
        super().__init__(*args, **kwargs)

    def do_GET(self):
        # Handle SSE endpoint
        if self.path == '/__live_reload_sse__':
            self.handle_sse()
            return

        # Serve files normally but inject script into HTML
        if self.path == '/':
            self.path = '/index.html'

        file_path = Path(os.getcwd()) / self.path.lstrip('/')

        # Security check
        try:
            file_path.resolve().relative_to(Path(os.getcwd()))
        except ValueError:
            self.send_error(403, "Forbidden")
            return

        if file_path.exists() and file_path.suffix == '.html':
            self.serve_html_with_injection(file_path)
        else:
            super().do_GET()

    def serve_html_with_injection(self, file_path):
        """Serve HTML with live reload script injected"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()

            # Inject live reload script before </body>
            modified_content = content.replace('</body>', f'{LIVE_RELOAD_SCRIPT}</body>')

            # Send response
            self.send_response(200)
            self.send_header('Content-Type', 'text/html; charset=utf-8')
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
            self.send_header('Content-Length', str(len(modified_content.encode())))
            self.end_headers()
            self.wfile.write(modified_content.encode())

        except Exception as e:
            self.send_error(500, f"Internal server error: {e}")

    def handle_sse(self):
        """Handle Server-Sent Events connection"""
        self.send_response(200)
        self.send_header('Content-Type', 'text/event-stream')
        self.send_header('Cache-Control', 'no-cache')
        self.send_header('Connection', 'keep-alive')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()

        # Send initial connection message
        self.wfile.write(b'data: {"action": "connected"}\n\n')
        self.wfile.flush()

        # Add to connected clients
        sse_clients.append(self)

        # Keep connection alive with periodic heartbeat
        try:
            while True:
                time.sleep(30)
                self.wfile.write(b':heartbeat\n\n')
                self.wfile.flush()
        except:
            # Client disconnected
            if self in sse_clients:
                sse_clients.remove(self)

    def log_message(self, format, *args):
        """Custom log format"""
        timestamp = datetime.now().strftime('%H:%M:%S')
        message = format % args

        # Color code based on status
        if "200" in message or "304" in message:
            # Success - green
            print(f"  [{timestamp}] ✓ {message}")
        elif "404" in message:
            # Not found - yellow
            print(f"  [{timestamp}] ⚠ {message}")
        elif "500" in message or "403" in message:
            # Error - red
            print(f"  [{timestamp}] ✗ {message}")
        else:
            # Info - default
            print(f"  [{timestamp}] → {message}")

def setup_file_watcher(server):
    """Set up file system watching"""
    observer = Observer()
    handler = FileChangeHandler(server)

    # Watch directories
    watch_paths = [
        ROOT_DIR,
        ROOT_DIR / 'css',
        ROOT_DIR / 'js',
        ROOT_DIR / 'js' / 'core',
        ROOT_DIR / 'js' / 'views',
        ROOT_DIR / 'data' / 'config'
    ]

    print("\n📁 Watching for changes in:")
    for path in watch_paths:
        if path.exists():
            observer.schedule(handler, str(path), recursive=False)
            print(f"  ✓ {path.relative_to(Path.cwd())}")
        else:
            print(f"  ⚠ {path.relative_to(Path.cwd())} (not found)")

    observer.start()
    return observer

def check_dependencies():
    """Check if required packages are installed"""
    try:
        import watchdog
        return True
    except ImportError:
        print("❌ Missing required package: watchdog")
        print("\nPlease install it using:")
        print("  pip install watchdog")
        print("\nOr use the basic server without live reload:")
        print("  ./serve.sh")
        return False

def main():
    """Main entry point"""

    if not check_dependencies():
        sys.exit(1)

    # Ensure frontend directory exists
    if not ROOT_DIR.exists():
        print(f"❌ Frontend directory not found: {ROOT_DIR}")
        sys.exit(1)

    print("\n🚀 VioBox Viewer Development Server with Live Reload")
    print("=" * 50)
    print(f"📡 Server URL:    http://localhost:{PORT}")
    print(f"📂 Serving from:  {ROOT_DIR.relative_to(Path.cwd())}")
    print(f"🔄 Live Reload:   Enabled")
    print(f"👀 Watching:      {', '.join(WATCH_EXTENSIONS)}")
    print("\nPress Ctrl+C to stop the server")

    # Create and start server
    try:
        server = HTTPServer(('', PORT), LiveReloadHTTPHandler)

        # Set up file watching in background thread
        observer = setup_file_watcher(server)

        # Open browser
        print("\n🌐 Opening browser...")
        webbrowser.open(f'http://localhost:{PORT}')

        print("\n📊 Server Activity:")
        print("-" * 50)

        # Start serving
        server.serve_forever()

    except KeyboardInterrupt:
        print("\n\n👋 Shutting down...")
        observer.stop()
        observer.join()
        server.shutdown()
        print("✅ Server stopped")
    except OSError as e:
        if 'Address already in use' in str(e):
            print(f"\n❌ Port {PORT} is already in use!")
            print("   Try one of these:")
            print(f"   1. Kill the process using port {PORT}")
            print("   2. Use a different port by editing PORT in this script")
        else:
            print(f"❌ Server error: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()